lib/src/characters_impl.dart

```diff
@@ -186,60 +188,96 @@ class StringCharacters extends Iterable<String> implements Characters {
+  ///
+  /// Starts from the index [cursor] in [string].
+  /// Use [breaks], which is assumed to be at [cursor],
+  /// if available.
+  ///
+  /// Returns `string.length` if there are less than [count]
+  /// characters left.
+  int _skipIndices(int count, int cursor, Breaks? breaks) {
+    if (count == 0 || cursor == string.length) return cursor;
+    breaks ??= Breaks(string, cursor, string.length, stateSoTNoBreak);
+    do {
+      var nextBreak = breaks.nextBreak();
+      if (nextBreak < 0) break;
+      cursor = nextBreak;
+    } while (--count > 0);
+    return cursor;
+  }
 
   @override
   Characters skip(int count) {
     RangeError.checkNotNegative(count, "count");
-    if (count == 0) return this;
-    if (string.isNotEmpty) {
-      int stringLength = string.length;
-      var breaks = Breaks(string, 0, stringLength, stateSoTNoBreak);
-      int startIndex = 0;
-      while (count > 0) {
-        int index = breaks.nextBreak();
-        if (index >= 0) {
-          count--;
-          startIndex = index;
-        } else {
-          return _empty;
-        }
-      }
-      if (startIndex == stringLength) return _empty;
-      return StringCharacters(string.substring(startIndex));
-    }
-    return this;
+    return _skip(count);
+  }
+
+  Characters _skip(int count) {
+    var start = _skipIndices(count, 0, null);
+    if (start == string.length) return _empty;
+    return StringCharacters(string.substring(start));
   }
 
   @override
   Characters take(int count) {
     RangeError.checkNotNegative(count, "count");
-    if (count == 0) return _empty;
-    if (string.isNotEmpty) {
-      var breaks = Breaks(string, 0, string.length, stateSoTNoBreak);
-      int endIndex = 0;
-      while (count > 0) {
-        int index = breaks.nextBreak();
-        if (index >= 0) {
-          endIndex = index;
-          count--;
-        } else {
-          return this;
+    return _take(count);
+  }
+
+  Characters _take(int count) {
+    var end = _skipIndices(count, 0, null);
+    if (end == string.length) return this;
+    return StringCharacters(string.substring(0, end));
+  }
+
+  @override
+  Characters getRange(int start, [int? end]) {
+    RangeError.checkNotNegative(start, "start");
+    if (end == null) return _skip(start);
+    if (end < start) throw RangeError.range(end, start, null, "end");
+    if (end == start) return _empty;
+    if (start == 0) return _take(end);
+    if (string.isEmpty) return this;
+    var breaks = Breaks(string, 0, string.length, stateSoTNoBreak);
+    var startIndex = _skipIndices(start, 0, breaks);
+    if (startIndex == string.length) return _empty;
+    var endIndex = _skipIndices(end - start, start, breaks);
+    return StringCharacters(string.substring(startIndex, endIndex));
+  }
+
+  @override
+  Characters characterAt(int position) {
+    var breaks = Breaks(string, 0, string.length, stateSoTNoBreak);
+    var start = 0;
+    findIndices:
```

Comment:
I am personally not a fan of using labels in dart - they are just really uncommon and thus can be confusing or require additional research to understand exactly what they do (I had no idea you could even use them to break out from a whole block like this).

Consider just inlining the two throw statements instead of doing `break findIndices;` which really just throws - or call a method that throws. That has a lot more straightforward/explicit behavior imo.

Reply:
With that attitude, there won't ever be common use of labels 😉 .

This code is probably just as good with inlined throws. Will do.